### ISSUE TITLE
MyReadingManga Extension has nonfunctioning manga covers after restoring a backup

### ISSUE THREAD

**Issue by Eminane** [inorichi/tachiyomi-extensions#1543]
After restoring a backup, I noticed that the manga covers from the MyReadingManga Extensions are missing.
![image](https://user-images.githubusercontent.com/55626138/65373903-85fe2d80-dc51-11e9-847c-84e3865b5f3b.png)
The manga covers from other sources loaded normally.
![image](https://user-images.githubusercontent.com/55626138/65373910-aded9100-dc51-11e9-8391-8fb6763e4223.png)
There was no change after restoring the backup twice more, restarting the phone twice, uninstalling and installing the extension, clearing chapter cache data, clearing cookies, and refreshing library metadata.
When viewing the MyReadingManga catalog, the manga that have not been added to the library yet had working covers. After adding the manga with working covers from the MyReadingManga Extension to the library, the covers continued to work.
![image](https://user-images.githubusercontent.com/55626138/65373973-8945e900-dc52-11e9-96b2-e41d3b2e8910.png)
I tried manually changing the cover but received an error message.
![image](https://user-images.githubusercontent.com/55626138/65373988-b72b2d80-dc52-11e9-972c-4dfae97a4796.png)

Also, unrelated to the cover errors, creating shortcuts to manga restored from the backup from the MyReadingManga Extension results in an error as well.
![image](https://user-images.githubusercontent.com/55626138/65374013-f9ed0580-dc52-11e9-982a-a8340fcb1ba0.png)

**Comment by joseph619** [comment]
@happywillow0

**Comment by happywillow0** [comment]
Hm, interesting. I've been working on the extension of late but I don't think have made any drastic changes on how things actually work. My changes don't include changes to the cover images. 

I remember having an issue a while back where the covers were failing to update. My solution at the time was to remove the items from my library, cleared the database (deleting manga not in my library) and adding it back in. 

I'm not totally sure what is included as part of the backup and restore process and what is stored in the database. I'm assuming you tried the "Refresh library metadata" option?

MRM did have a period of time where the images were lost and they had to re-upload but I'm not sure if that affects anything with the app's database. 

I'm going to test the backup and restore on one of my other phones.

**Comment by happywillow0** [comment]
Also, I remember and found again this open issue on the main app side about error setting covers which may or may not be related.
https://github.com/inorichi/tachiyomi/issues/1663

**Comment by Eminane** [comment]
> Hm, interesting. I've been working on the extension of late but I don't think have made any drastic changes on how things actually work. My changes don't include changes to the cover images.
> 
> I remember having an issue a while back where the covers were failing to update. My solution at the time was to remove the items from my library, cleared the database (deleting manga not in my library) and adding it back in.
> 
> I'm not totally sure what is included as part of the backup and restore process and what is stored in the database. I'm assuming you tried the "Refresh library metadata" option?
> 
> MRM did have a period of time where the images were lost and they had to re-upload but I'm not sure if that affects anything with the app's database.
> 
> I'm going to test the backup and restore on one of my other phones.

I read the MRM issue from 2018 prior to reporting my issue. My problem is specific to backups and restores, so I think the issues are separate. 

I've tried "Refresh library metadata" several times, each time in a different order with restarting the app, rebooting the phone, clearing cache and cookies, and uninstalling and reinstalling the extension.
 
Honestly, I would rather keep the blank covers than removing all the items from the library and adding them back on. I have way too many titles to go through the trouble of doing that...

I hope this gets resolved, 
thanks!

**Comment by happywillow0** [comment]
You had to make me think about this. I was typing something else and I figured out what the issue was. 
I'm halfway to a fix but I have to go to work soon so I may not be able to finish.

**Comment by Eminane** [comment]
> You had to make me think about this. I was typing something else and I figured out what the issue was.
> I'm halfway to a fix but I have to go to work soon so I may not be able to finish.

I'm so sorry for the trouble!! I'm not in a rush to have this fixed, no pressure!
Thank you so much though, I really appreciate it!!

**Comment by happywillow0** [comment]
I'm at a roadblock. Here is what is happening:

When you browse the main extension to see what is new, it is pulling information off the home page. You can get the title, the url, and the thumbnail and everything is pretty. 

When you actually go into the details, screen, more data is pulled such as the description, the genres, status, etc. The app has an option to pull a thumbnail again from the manga page. If the manga page thumbnail is omitted, Tachiyomi uses the first thumbnail it found as part of browse. 

On the backup and restore is that the manga creation process, the following data is saved:
`{"manga":["/yamamoto-kotetsuko-o-mairi-desu-yo-blessing-house-eng/"," Omairi Desu yo ~ A Blessing On Your House! ",6318378620301107421,0,0]` plus chapter read status and sync data. 
Because the title and url is already known, Tachiyomi is skipping to the details parse section. 

Comparing the parse section of MRM with another extension such as MangaHere, this is where I noticed the optional second thumbnail option. Because the thumbnail is not found during the details parse, it isn't there and comes up blank. 

There is a separate issue where you are unable to update a thumbnail because when it doesn't exist (https://github.com/inorichi/tachiyomi/issues/1663)

So, finally, onto the conundrum. We can add the second thumbnail pull in. However, the main manga page for MRM doesn't include a thumbnail anymore so we have to pull one of the images. Some manga are ok where the 1st image is the cover but for some other manga, the 1st page is credits. 

tldr: How do I pull the proper cover page off the the main manga's page where I don't know which image is the cover itself?

**Comment by Eminane** [comment]
> I'm at a roadblock. Here is what is happening:
> 
> When you browse the main extension to see what is new, it is pulling information off the home page. You can get the title, the url, and the thumbnail and everything is pretty.
> 
> When you actually go into the details, screen, more data is pulled such as the description, the genres, status, etc. The app has an option to pull a thumbnail again from the manga page. If the manga page thumbnail is omitted, Tachiyomi uses the first thumbnail it found as part of browse.
> 
> On the backup and restore is that the manga creation process, the following data is saved:
> `{"manga":["/yamamoto-kotetsuko-o-mairi-desu-yo-blessing-house-eng/"," Omairi Desu yo ~ A Blessing On Your House! ",6318378620301107421,0,0]` plus chapter read status and sync data.
> Because the title and url is already known, Tachiyomi is skipping to the details parse section.
> 
> Comparing the parse section of MRM with another extension such as MangaHere, this is where I noticed the optional second thumbnail option. Because the thumbnail is not found during the details parse, it isn't there and comes up blank.
> 
> There is a separate issue where you are unable to update a thumbnail because when it doesn't exist ([inorichi/tachiyomi#1663](https://github.com/inorichi/tachiyomi/issues/1663))
> 
> So, finally, onto the conundrum. We can add the second thumbnail pull in. However, the main manga page for MRM doesn't include a thumbnail anymore so we have to pull one of the images. Some manga are ok where the 1st image is the cover but for some other manga, the 1st page is credits.
> 
> tldr: How do I pull the proper cover page off the the main manga's page where I don't know which image is the cover itself?

I'll apologize in advanced if I sound like I don't know what I'm talking about. (I'm only recently learning coding C#)

So the backup feature for MRM doesn't include the thumbnail, and that's why nothing's showing up (is what I'm gathering). 
I also don't have a super clear understanding of how Tachiyomi works. In this case, once the restoring retrieves the manga information for MRM, that manga just won't have the thumbnail information while all other manga does have it?
Is this something that I need to fix for each manga, or can you fix the code to restore the original thumbnails?
If the simplest way is to get the first page as the cover, that's probably better than not seeing anything.

**Comment by SnakeDoc83** [comment]
>tldr: How do I pull the proper cover page off the the main manga's page where I don't know which image is the cover itself?

Like Eminane said, select the first image `document.select("img.img-myreadingmanga").first().attr("src")` and use that.  Sometimes you'll have the right image, other times you won't, but it's still something.

**Comment by joseph619** [comment]
I agree, a wrong cover is still better than no cover.

**Comment by Eminane** [comment]
> > tldr: How do I pull the proper cover page off the the main manga's page where I don't know which image is the cover itself?
> 
> Like Eminane said, select the first image `document.select("img.img-myreadingmanga").first().attr("src")` and use that. Sometimes you'll have the right image, other times you won't, but it's still something.

Well, I mean... I prefer to have the right cover. But if that means that I have to go through each individual manga to do (and I have over 1000, not kidding), then I'd rather do the quickest method even if it gets the majority of the covers wrong.

**Comment by happywillow0** [comment]
> So the backup feature for MRM doesn't include the thumbnail, and that's why nothing's showing up (is what I'm gathering)
Rather, the backup tells the Tachiyomi where to get the information, however the section getting the information in the extension doesn't include a thumbnail. Once I add it in, refreshing data should pull properly or at least starting over with a new restore. 

I have the code built in Android Studio at home so it is pretty much ready to go. I'm just wanted to double check if I can do it more elegantly.

I posted it to Discord but was wondering, can I detect if `SManga.thumbnail_url ` already exists or not? I want to see if we can make it so only `if (thumbnail_url = null) {load first page as thumbnail]`
You guys think this is possible?

**Comment by Eminane** [comment]
> I posted it to Discord but was wondering, can I detect if `SManga.thumbnail_url ` already exists or not? I want to see if we can make it so only `if (thumbnail_url = null) {load first page as thumbnail]`
> You guys think this is possible?

I aspire to understand what any of this actually means, but I understand that things are looking good (I think).
Not that my judgement weighs any amount in this, but I think that looks functional...

**Comment by happywillow0** [comment]
No worries. I self taught myself everything I have worked on so far. 

I submitted the PR for the fix I threw together. 

Because I didn't want to break the cover functionality when adding new mangas to the library, I managed add some basic conditions. If approved, the covers will be set to the 1st page. 

Since the cover cache is already broken due to the blank covers, I think you may have to start over and clear the app data. Once the library and cache is clear, you can restore the backup and it should load everything fresh. 

Some more info on the conditions, think of this when you first open the app after you close it / force stop:

- After first load, covers in the library will be set to the first page of the manga if requested via a restore or refresh. 
- If you open the browse section from the catalog, this function stops and the cover will be set properly (if you are adding a new manga)

**Comment by Eminane** [comment]
> Because I didn't want to break the cover functionality when adding new mangas to the library, I managed add some basic conditions. If approved, the covers will be set to the 1st page.
It's a shame it's not the right page, but I really appreciate you helping me with this.

> Since the cover cache is already broken due to the blank covers, I think you may have to start over and clear the app data. Once the library and cache is clear, you can restore the backup and it should load everything fresh.
So once I do that, MRM will load the first page as the cover? 

> Some more info on the conditions, think of this when you first open the app after you close it / force stop:
> 
> * After first load, covers in the library will be set to the first page of the manga if requested via a restore or refresh.
> * If you open the browse section from the catalog, this function stops and the cover will be set properly (if you are adding a new manga)

So new manga will have the right cover but this one's I already have will have the first page in the catalog. Out of curiosity, will this be fully fixed in the future? I'm assuming the first page thing is a temp fix.
Also, if your fix is approved, will there be an extension update?
Thanks for everything!!
